codex-rs/core/src/environment_context.rs

```diff
@@ -71,6 +72,29 @@ impl EnvironmentContext {
             shell,
         }
     }
+
+    /// Compares two environment contexts, ignoring the shell. Useful when
+    /// comparing turn to turn, since the initial environment_context will
+    /// include the shell, and then it is not configurable from turn to turn.
+    pub fn equals_except_shell(&self, other: &EnvironmentContext) -> bool {
+        self.cwd == other.cwd
+            && self.approval_policy == other.approval_policy
+            && self.sandbox_mode == other.sandbox_mode
+            && self.network_access == other.network_access
+            && self.writable_roots == other.writable_roots
+    }
```

Comment:
I think if you destructure the fields of self and other then you can force yourself to handle all fields and `_` shell. You don't want to add a new field and forget to update this.